File: ios/Signalling.m
//
//  Signalling.m
//  iOS.Conference.WebRTC
//
//  Created by Adrian Mercado on 2014-06-09.
//  Copyright (c) 2014 Frozen Mountain Software. All rights reserved.
//

#import "Signalling.h"

@implementation Signalling

// IceLink includes a WebSync client extension that will
// automatically manage signalling for you. If you are not
// using WebSync, set this to false to see how the event
// system works. Use it as a template for your own code.
static bool useWebSyncExtension = YES;

- (id)initWithWebSyncServerUrl:(NSString *)websyncServerUrl
{
  self = [super init];
  if(self)
  {
    _websyncServerUrl = [websyncServerUrl copy];

    _sendOfferAnswerEvent = [FMCallback callback:@selector(sendOfferAnswer:) target:self];
    _sendCandidateEvent = [FMCallback callback:@selector(sendCandidate:) target:self];
    _receiveOfferAnswerOrCandidateEvent = [FMCallback callback:@selector(receiveOfferAnswerOrCandidate:) target:self];
  }
  return self;
}

- (void)start:(void (^)(NSString *))callback
{
  // Create a WebSync client.
  _websyncClient = [[FMWebSyncClient alloc] initWithRequestUrl:_websyncServerUrl];
  //[_websyncClient setDomainKey:[FMGuid guidWithG:@"5fb3bdc2-ea34-11dd-9b91-3e6b56d89593"]]; // WebSync On-Demand

  FMWebSyncConnectArgs *connectArgs = [FMWebSyncConnectArgs connectArgs];
  [connectArgs setOnFailureBlock:^(FMWebSyncConnectFailureArgs *e)
   {
     callback([NSString stringWithFormat:@"Could not connect to WebSync. %@", e.exception.message]);
     e.retry = NO;
   }];
  [connectArgs setOnSuccessBlock:^(FMWebSyncConnectSuccessArgs *e)
   {
     callback(nil);
   }];
  [_websyncClient connectWithArgs:connectArgs];
}

- (void)stop:(void (^)(NSString *))callback
{
  FMWebSyncDisconnectArgs *disconnectArgs = [FMWebSyncDisconnectArgs disconnectArgs];
  [disconnectArgs setOnCompleteBlock:^(FMWebSyncDisconnectCompleteArgs *e)
   {
     callback(nil);
   }];
  [_websyncClient disconnectWithArgs:disconnectArgs];
  _websyncClient = nil;
}

- (void)attach:(FMIceLinkConference *)conference sessionId:(NSString *)sessionId callback:(void (^)(NSString *))callback
{
  _conference = conference;
  _sessionId = [sessionId copy];

  NSString *channel = [NSString stringWithFormat:@"/%@", _sessionId];
  if (useWebSyncExtension)
  {
    // Manage the conference automatically using a WebSync
    // channel. P2P links will be created automatically to
    // peers that join the same channel.
    FMIceLinkWebSyncJoinConferenceArgs *joinArgs = [FMIceLinkWebSyncJoinConferenceArgs joinConferenceArgsWithConferenceChannel:channel conference:_conference];
    [joinArgs setOnFailureBlock:^(FMIceLinkWebSyncJoinConferenceFailureArgs *e)
     {
       callback([NSString stringWithFormat:@"Could not attach signalling to conference %@. %@", _sessionId, e.exception.message]);
     }];
    [joinArgs setOnSuccessBlock:^(FMIceLinkWebSyncJoinConferenceSuccessArgs *e)
     {
       callback(nil);
     }];
    [_websyncClient joinConferenceWithArgs:joinArgs];
  }
  else
  {
    // When the conference generates an offer/answer or candidate,
    // we want to send it to the remote peer immediately.
    [_conference addOnLinkOfferAnswer:_sendOfferAnswerEvent];
    [_conference addOnLinkCandidate:_sendCandidateEvent];

    // When we receive an offer/answer or candidate, we want to
    // inform the conference immediately.
    [_websyncClient addOnNotify:_receiveOfferAnswerOrCandidateEvent];

    // Subscribe to a WebSync channel. When another client joins the same
    // channel, create a P2P link. When a client leaves, destroy it.
    FMWebSyncSubscribeArgs *subscribeArgs = [FMWebSyncSubscribeArgs subscribeArgsWithChannel:channel];
    [subscribeArgs setOnFailureBlock:^(FMWebSyncSubscribeFailureArgs *e)
     {
       callback([NSString stringWithFormat:@"Could not attach signalling to conference %@. %@", _sessionId, e.exception.message]);
     }];
    [subscribeArgs setOnReceiveBlock:^(FMWebSyncSubscribeReceiveArgs *e) { }];
    [subscribeArgs setOnSuccessBlock:^(FMWebSyncSubscribeSuccessArgs *e)
     {
       callback(nil);
     }];
    [subscribeArgs setOnClientSubscribeBlock:^(FMWebSyncSubscribersClientSubscribeArgs *e)
     {
       // Kick off a P2P link.
       NSString *peerId = [e.subscribedClient.clientId toString];
       NSObject *peerState = e.subscribedClient.boundRecords;
       [_conference linkWithPeerId:peerId peerState:peerState];
     }];
    [subscribeArgs setOnClientUnsubscribeBlock:^(FMWebSyncSubscribersClientUnsubscribeArgs *e)
     {
       // Tear down a P2P link.
       NSString *peerId = [e.unsubscribedClient.clientId toString];
       [_conference unlinkWithPeerId:peerId];
     }];
    [_websyncClient subscribeWithArgs:subscribeArgs];
  }
}

- (void)sendOfferAnswer:(FMIceLinkLinkOfferAnswerArgs *)e
{
  NSString *tag = [NSString stringWithFormat:@"offeranswer:%@", _sessionId];
  [_websyncClient notifyWithArgs:[FMWebSyncNotifyArgs notifyArgsWithClientId:[FMGuid guidWithG:e.peerId]
                                                                    dataJson:[e.offerAnswer toJson]
                                                                         tag:tag]];
}

- (void)sendCandidate:(FMIceLinkLinkCandidateArgs *)e
{
  NSString *tag = [NSString stringWithFormat:@"candidate:%@", _sessionId];
  [_websyncClient notifyWithArgs:[FMWebSyncNotifyArgs notifyArgsWithClientId:[FMGuid guidWithG:e.peerId]
                                                                    dataJson:[e.candidate toJson]
                                                                         tag:tag]];
}

- (void)receiveOfferAnswerOrCandidate:(FMWebSyncNotifyReceiveArgs *)e
{
  NSString *peerId = [e.notifyingClient.clientId toString];
  NSObject *peerState = e.notifyingClient.boundRecords;
  NSString *offerAnswerTag = [NSString stringWithFormat:@"offeranswer:%@", _sessionId];
  NSString *candidateTag = [NSString stringWithFormat:@"candidate:%@", _sessionId];
  if ([e.tag isEqualToString:offerAnswerTag])
  {
    [_conference receiveOfferAnswer:[FMIceLinkOfferAnswer fromJsonWithOfferAnswerJson:e.dataJson]
                             peerId:peerId
                          peerState:peerState];
  }
  else if ([e.tag isEqualToString:candidateTag])
  {
    [_conference receiveCandidate:[FMIceLinkCandidate fromJsonWithCandidateJson:e.dataJson]
                           peerId:peerId];
  }
}

- (void)detach:(void (^)(NSString *))callback
{
  NSString *channel = [NSString stringWithFormat:@"/%@", _sessionId];
  if (useWebSyncExtension)
  {
    // Leave the managed WebSync channel.
    FMIceLinkWebSyncLeaveConferenceArgs *leaveArgs = [FMIceLinkWebSyncLeaveConferenceArgs leaveConferenceArgsWithConferenceChannel:channel];
    [leaveArgs setOnFailureBlock:^(FMIceLinkWebSyncLeaveConferenceFailureArgs *e)
     {
       callback([NSString stringWithFormat:@"Could not detach signalling from conference %@. %@", _sessionId, e.exception.message]);
     }];
    [leaveArgs setOnSuccessBlock:^(FMIceLinkWebSyncLeaveConferenceSuccessArgs *e)
     {
       _conference = nil;
       _sessionId = nil;
       callback(nil);
     }];
    [_websyncClient leaveConferenceWithArgs:leaveArgs];
  }
  else
  {
    // Unsubscribe from the WebSync channel.
    FMWebSyncUnsubscribeArgs *unsubscribeArgs = [FMWebSyncUnsubscribeArgs unsubscribeArgsWithChannel:channel];
    [unsubscribeArgs setOnFailureBlock:^(FMWebSyncUnsubscribeFailureArgs *e)
     {
       callback([NSString stringWithFormat:@"Could not detach signalling from conference %@. %@", _sessionId, e.exception.message]);
     }];
    [unsubscribeArgs setOnSuccessBlock:^(FMWebSyncUnsubscribeSuccessArgs *e)
     {
       // Detach our event handlers.
       [_conference removeOnLinkOfferAnswer:_sendOfferAnswerEvent];
       [_conference removeOnLinkCandidate:_sendCandidateEvent];
       [_websyncClient removeOnNotify:_receiveOfferAnswerOrCandidateEvent];

       _conference = nil;
       _sessionId = nil;
       callback(nil);
     }];
    [_websyncClient unsubscribeWithArgs:unsubscribeArgs];
  }
}

@end

File: Android.Conference.WebRTC/app/src/main/java/fm/android/conference/webrtc/Signalling.java
package fm.android.conference.webrtc;

import java.util.Locale;

import fm.*;
import fm.icelink.*;
import fm.icelink.websync.*;
import fm.websync.*;
import fm.websync.subscribers.*;

// Peers have to exchange information when setting up P2P links,
// like descriptions of the streams (called the offer or answer)
// and information about network addresses (called candidates).
// IceLink generates this information for you automatically.
// Your responsibility is to pass messages back and forth between
// peers as quickly as possible. This is called "signalling".
public class Signalling
{
    // We're going to use WebSync for this example, but any real-time
    // messaging system will do (like SIP or XMPP). We use WebSync
    // since it works well with JavaScript and uses HTTP, which is
    // widely allowed. To use something else, simply replace the calls
    // to WebSync with calls to your library.
    private String websyncServerUrl = null;
    private Client websyncClient = null;

    // IceLink includes a WebSync client extension that will
    // automatically manage signalling for you. If you are not
    // using WebSync, set this to false to see how the event
    // system works. Use it as a template for your own code.
    private boolean useWebSyncExtension = true;
    
    private SingleAction<LinkOfferAnswerArgs> sendOfferAnswerEvent;
    private SingleAction<LinkCandidateArgs>   sendCandidateEvent;
    private SingleAction<NotifyReceiveArgs>   receiveOfferAnswerOrCandidateEvent;
    
    public Signalling(String websyncServerUrl)
    {
        this.websyncServerUrl = websyncServerUrl;

        sendOfferAnswerEvent = new SingleAction<LinkOfferAnswerArgs>() { public void invoke(LinkOfferAnswerArgs e) { sendOfferAnswer(e); } };
        sendCandidateEvent   = new SingleAction<LinkCandidateArgs>()   { public void invoke(LinkCandidateArgs e)   { sendCandidate(e);   } };
        receiveOfferAnswerOrCandidateEvent = new SingleAction<NotifyReceiveArgs>() { public void invoke(NotifyReceiveArgs e) { receiveOfferAnswerOrCandidate(e); } };
    }
    
    public void start(final SingleAction<String> callback) throws Exception
    {
        // Create a WebSync client.
        websyncClient = new Client(websyncServerUrl);
        //websyncClient.setDomainKey(new Guid("5fb3bdc2-ea34-11dd-9b91-3e6b56d89593")); // WebSync On-Demand

        // Create a persistent connection to the server.
        websyncClient.connect(new ConnectArgs()
        {{
            setOnFailure(new SingleAction<ConnectFailureArgs>()
            {
                public void invoke(ConnectFailureArgs e)
                {
                    callback.invoke(String.format(Locale.getDefault(), "Could not connect to WebSync. %s", e.getException().getMessage()));
                    e.setRetry(false);
                }
            });
            setOnSuccess(new SingleAction<ConnectSuccessArgs>()
            {
                public void invoke(ConnectSuccessArgs e)
                {
                    callback.invoke(null);
                }
            });
        }});
    }

    public void stop(final SingleAction<String> callback) throws Exception
    {
        // Tear down the persistent connection.
        websyncClient.disconnect(new DisconnectArgs()
        {{
            setOnComplete(new SingleAction<DisconnectCompleteArgs>()
            {
                public void invoke(DisconnectCompleteArgs e)
                {
                    callback.invoke(null);
                }
            });
        }});

        websyncClient = null;
    }

    private Conference conference = null;
    private String sessionId = null;

    public void attach(final Conference conference, final String sessionId, final SingleAction<String> callback) throws Exception
    {
        this.conference = conference;
        this.sessionId = sessionId;

        if (useWebSyncExtension)
        {
            // Manage the conference automatically using a WebSync
            // channel. P2P links will be created automatically to
            // peers that join the same channel.
            ClientExtensions.joinConference(websyncClient, new JoinConferenceArgs("/" + sessionId, conference)
            {{
                setOnFailure(new SingleAction<JoinConferenceFailureArgs>()
                {
                    public void invoke(JoinConferenceFailureArgs e)
                    {
                        callback.invoke(String.format(Locale.getDefault(), "Could not attach signalling to conference %s. %s", sessionId, e.getException().getMessage()));
                    }
                });
                setOnSuccess(new SingleAction<JoinConferenceSuccessArgs>()
                {
                    public void invoke(JoinConferenceSuccessArgs e)
                    {
                        callback.invoke(null);
                    }
                });
            }});
        }
        else
        {
            // When the conference generates an offer/answer or candidate,
            // we want to send it to the remote peer immediately.
            conference.addOnLinkOfferAnswer(sendOfferAnswerEvent);
            conference.addOnLinkCandidate(sendCandidateEvent);

            // When we receive an offer/answer or candidate, we want to
            // inform the conference immediately.
            websyncClient.addOnNotify(receiveOfferAnswerOrCandidateEvent);

            // Subscribe to a WebSync channel. When another client joins the same
            // channel, create a P2P link. When a client leaves, destroy it.

            // Subscribe to a WebSync channel. When another client joins the same
            // channel, create a P2P link. When a client leaves, destroy it.
            SubscribeArgs subscribeArgs = new SubscribeArgs("/" + sessionId)
            {{
                setOnFailure(new SingleAction<SubscribeFailureArgs>()
                {
                    public void invoke(SubscribeFailureArgs e)
                    {
                        callback.invoke(String.format(Locale.getDefault(), "Could not attach signalling to conference %s. %s", sessionId, e.getException().getMessage()));
                    }
                });
                setOnSuccess(new SingleAction<SubscribeSuccessArgs>()
                {
                    public void invoke(SubscribeSuccessArgs e)
                    {
                        callback.invoke(null);
                    }
                });
                setOnReceive(new SingleAction<SubscribeReceiveArgs>() { public void invoke(SubscribeReceiveArgs e) { } });
            }};
            SubscribeArgsExtensions.setOnClientSubscribe(subscribeArgs, new SingleAction<ClientSubscribeArgs>()
            {
                public void invoke(ClientSubscribeArgs e)
                {
                    try
                    {
                        // Kick off a P2P link.
                        String peerId = e.getSubscribedClient().getClientId().toString();
                        Object peerState = e.getSubscribedClient().getBoundRecords();
                        conference.link(peerId, peerState);
                    }
                    catch (Exception ex)
                    { ex.printStackTrace(); }
                }
            });
            SubscribeArgsExtensions.setOnClientUnsubscribe(subscribeArgs, new SingleAction<ClientUnsubscribeArgs>()
            {
                public void invoke(ClientUnsubscribeArgs e)
                {
                    try
                    {
                        // Tear down a P2P link.
                        String peerId = e.getUnsubscribedClient().getClientId().toString();
                        conference.unlink(peerId);
                    }
                    catch (Exception ex)
                    { ex.printStackTrace(); }
                }
            });
            websyncClient.subscribe(subscribeArgs);
        }
    }

    private void sendOfferAnswer(LinkOfferAnswerArgs e)
    {
        try
        {
            websyncClient.notify(new NotifyArgs(new Guid(e.getPeerId()), e.getOfferAnswer().toJson(), "offeranswer:" + sessionId));
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
    }

    private void sendCandidate(LinkCandidateArgs e)
    {
        try
        {
            websyncClient.notify(new NotifyArgs(new Guid(e.getPeerId()), e.getCandidate().toJson(), "candidate:" + sessionId));
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
    }

    private void receiveOfferAnswerOrCandidate(NotifyReceiveArgs e)
    {
        try
        {
            String peerId = e.getNotifyingClient().getClientId().toString();
            Object peerState = e.getNotifyingClient().getBoundRecords();
            if (e.getTag().equals("offeranswer:" + sessionId))
            {
                conference.receiveOfferAnswer(OfferAnswer.fromJson(e.getDataJson()), peerId, peerState);
            }
            else if (e.getTag().equals("candidate:" + sessionId))
            {
                conference.receiveCandidate(Candidate.fromJson(e.getDataJson()), peerId);
            }
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
    }

    public void detach(final SingleAction<String> callback) throws Exception
    {
        if (useWebSyncExtension)
        {
            // Leave the managed WebSync channel.
            ClientExtensions.leaveConference(websyncClient, new LeaveConferenceArgs("/" + sessionId)
            {{
                setOnSuccess(new SingleAction<LeaveConferenceSuccessArgs>()
                {
                    public void invoke(LeaveConferenceSuccessArgs e)
                    {
                        conference = null;
                        sessionId = null;
                        
                        callback.invoke(null);
                    }
                });
                setOnFailure(new SingleAction<LeaveConferenceFailureArgs>()
                {
                    public void invoke(LeaveConferenceFailureArgs e)
                    {
                        callback.invoke(String.format(Locale.getDefault(), "Could not detach signalling from conference %s. %s", sessionId, e.getException().getMessage()));
                    }
                });
            }});
        }
        else
        {
            // Unsubscribe from the WebSync channel.
            websyncClient.unsubscribe(new UnsubscribeArgs("/" + sessionId)
            {{
                setOnSuccess(new SingleAction<UnsubscribeSuccessArgs>()
                {
                    public void invoke(UnsubscribeSuccessArgs e)
                    {
                        // Detach our event handlers.
                        conference.removeOnLinkOfferAnswer(sendOfferAnswerEvent);
                        conference.removeOnLinkCandidate(sendCandidateEvent);
                        websyncClient.removeOnNotify(receiveOfferAnswerOrCandidateEvent);

                        conference = null;
                        sessionId = null;
                        
                        callback.invoke(null);
                    }
                });
                setOnFailure(new SingleAction<UnsubscribeFailureArgs>()
                {
                    public void invoke(UnsubscribeFailureArgs e)
                    {
                        callback.invoke(String.format(Locale.getDefault(), "Could not detach signalling from conference %s. %s", sessionId, e.getException().getMessage()));
                    }
                });
            }});
        }
    }
}
